ocaml/xapi-guard/lib/disk_cache.ml

```diff
@@ -585,19 +567,29 @@ end = struct
     let now = Mtime_clock.now () in
     let root = cache_of typ in
     let* contents = get_all_contents root in
-    let* () =
-      contents
-      |> List.map (get_fs_action root now)
-      |> Lwt_list.iter_p (commit __FUNCTION__)
+    let actions, pending =
+      contents |> List.map (get_fs_action root now) |> List.split
     in
-    delete_empty_dirs ~delete_root:false root
+    let* () = Lwt_list.iter_p (commit __FUNCTION__) actions in
+    let* () = delete_empty_dirs ~delete_root:false root in
+    Lwt.return (List.filter_map Fun.id pending)
 end
 
 let setup typ read write =
-  let* () = Setup.retime_cache_contents typ in
-  let queue, push = Lwt_bounded_stream.create 2 in
+  let* pending = Setup.retime_cache_contents typ in
+  let capacity = 4096 in
+  let queue, push = Lwt_bounded_stream.create capacity in
   let lock = Lwt_mutex.create () in
-  let q = {queue; push; lock; state= Disengaged} in
+  let state =
+    if pending = [] then
+      Direct
+    else if List.length pending < capacity then
```

Comment:
Here we are using non-tail recursive functions on a potentially large list. I believe 4096 is not causing a stack overflow but I do think this large enough to wonder if this is the right data structure.

Reply:
I'll try using sequences here, it's a bit involved since the lwt libraries like to use lists

Reply:
I've changed this to an tail-recursive iter

Reply:
Did you? I see a map below.